File: flutter/lib/features/champion/data/datasource/patch_version_local_data_source.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:lol_champion/core/constants/app_constants.dart';

class PatchVersionLocalDataSource {
  const PatchVersionLocalDataSource(this._storage);

  static const _storageKey = 'latest_lol_patch_version';

  final FlutterSecureStorage _storage;

  Future<String> read() async {
    try {
      final version = await _storage.read(key: _storageKey);
      return version == null || version.trim().isEmpty
          ? AppConstants.fallbackPatchVersion
          : version;
    } on Object {
      return AppConstants.fallbackPatchVersion;
    }
  }

  Future<void> write(String version) async {
    if (version.trim().isEmpty) {
      return;
    }
    await _storage.write(key: _storageKey, value: version);
  }
}

File: react/src/features/champion/data/data-sources/patch-version-local-data-source.ts
import type { KeyValueStorage } from '@/core/storage/key-value-storage';

const patchVersionStorageKey = 'latest_lol_patch_version';

export interface PatchVersionLocalDataSource {
  readVersion(): string | null;
  writeVersion(version: string): void;
}

export class BrowserPatchVersionLocalDataSource implements PatchVersionLocalDataSource {
  constructor(private readonly storage: KeyValueStorage) {}

  readVersion(): string | null {
    return this.storage.get(patchVersionStorageKey);
  }

  writeVersion(version: string): void {
    this.storage.set(patchVersionStorageKey, version);
  }
}
